numpy/core/src/multiarray/methods.c

```diff
@@ -823,6 +825,47 @@ array_astype(PyArrayObject *self, PyObject *args, PyObject *kwds)
             return NULL;
         }
 
+        /*
+         * after casting from NPY_DATETIME_NAT to float64, we go back
+         * and replace the spurious floating point values with NPY_NAN
+         * at the appropriate indices; this may also be possible to
+         * achieve upstream, but the PyArray_CopyInto() operation above
+         * uses very low level general array iteration loops where it
+         * may not be desirable to perform a check for NAT on each
+         * iteration
+         */
+
+        /* confine the NaT to NaN substitution checks by type cast */
+        if ((PyArray_DESCR(self)->type_num == NPY_TIMEDELTA) &&
+            ((PyArray_DESCR(ret)->type_num == NPY_HALF)   ||
+            (PyArray_DESCR(ret)->type_num == NPY_FLOAT)   ||
+            (PyArray_DESCR(ret)->type_num == NPY_DOUBLE)  ||
+            (PyArray_DESCR(ret)->type_num == NPY_LONGDOUBLE))) {
+
+            PyObject *item;
+            PyArrayIterObject *self_iter;
+            PyObject *ret_iter;
+
+            /* array iterator objects */
+            self_iter = (PyArrayIterObject *)PyArray_IterNew((PyObject *)self);
+            ret_iter = PyArray_IterNew((PyObject *)ret);
+
+            while (self_iter->index < self_iter->size) {
+               /* GETITEM will retrieve Py_None for NaT */
+               item = PyArray_GETITEM(self, PyArray_ITER_DATA(self_iter));
+               Py_XDECREF(item);
```

Comment:
This ought to be after you're done with it, not before (although strictly it doesn't matter)